peps/pep-0795.rst

```diff
@@ -0,0 +1,2333 @@
+exacerbates this situation as ``x.bar(42)`` may now fail because ``x``
+has been made immutable. However, in contrast to failures due to
+reflective changes of a class, a ``NotFreezableError`` will point to the
+place in the code where the object was made immutable. This should facilitate
+debugging.
+
+It is possible that there are aliases of an object that is being
+frozen. Thus the type change will lead to exceptions if those
+aliases are accessed in a mutating way. This is the same as the
+current situation in Python where an object may be mutated in
+unexpected ways that can affect the objects type.
+
+Making freeze a soft keyword as opposed to a function has been proposed
+to facilitate flow typing. We believe this is an excellent proposal to
+consider for the future in conjunction with work on typing immutability.
+
+
+Data-race free Python
+======================
+
+This PEP adds support for immutable objects to Python and importantly
+permits sub-interpreters to directly share *immutable* objects. As a
+future PEP will we propose an extension that also permits directly
+sharing *mutable* objects. We refer to this proposal as “data-race free
+Python” or “DRF Python” for short. In that future proposal, mutable
+objects will only be accessible to one sub-interpreter at a time, but
+can be passed around and shared by reference. We believe that making
+Python data-race free is the right programming model for the langauge,
+and aim to be compatible with both sub-interpreters and free-threaded
+Python.
+
+DRF Python will borrow concepts from ownership (namely region-based
+ownership, see e.g. `Cyclone <https://cyclone.thelanguage.org/>`__) to
+make Python programs data-race free by construction. A description of
+the ownership model that we envision can be found in our PLDI 2025:
+`Dynamic Region Ownership for Concurrency
+Safety <https://wrigstad.com/pldi2025.pdf>`__.
+
+It is important to point out that data-race free Python is different
+from free-threaded Python, but aims to be fully compatible with it, and
+we believe that they can benefit from each other. In essence
+free-threaded Python focuses on making the CPython run-time resilient
+against data races in Python programs: a poorly synchronized Python
+program should not be able to corrupt reference counts, or other parts
+of the Python interpreter. The complementary goal pursued by DRF Python
+is to make it impossible for Python programs to have data races. Support
+for deeply immutable data can be seen as the first step towards this
+goal.
+
+
+Reference implementation
+########################
+
+Our `reference implementation`_, which we have made publicly available on
+Github, implements all of the features described above.
+We steal a bit from the top of the reference count in the object header. In 64-bit
+python this comes from the (currently unused) top word and thus has no impact at
+all on the runtime. In 32-bit python, the second bit of the word is used (since negative
```

Comment:
```suggestion
Python this comes from the (currently unused) top word and thus has no impact at
all on the runtime. In 32-bit Python, the second bit of the word is used (since negative
```